### PR TITLE
Reworked counsel-fzf to support ctl-u and prompt

### LINKED ISSUE #1280: counsel-fzf should support C-u to select the starting directory
Similar to, say, counsel-ag/rg it would be nice if counsel-fzf allowed to choose the start directory, so the setq to the initial directory in the defun counsel-fzf function could be

```
  (setq counsel--fzf-dir (if current-prefix-arg
                            (read-directory-name (concat
                                                  (car (split-string counsel-fzf-cmd))
                                                  " in directory: "))
                          (if
                              (and
                               (fboundp 'projectile-project-p)
```

to have the directory be chosen if prefix arg is passed? Since this is only a 3-lines change I figured an issue would be enough vs a pull but let me know.

### PR DESCRIPTION
**Pull request by woodensquares**
Resolves #1280 

Besides ctrl-u I have also added support for specifying a custom prompt similarly to the other counsel- functions.

### CODE REVIEW

**Comment by abo-abo** [comment]
Thanks, since the change is above the 15 lines, it needs an Emacs Copyright Assignment.  If you're willing to get it, see the README for more info.

**Comment by woodensquares** [comment]
I should already have a copyright assignment on file, I sent you an email to discuss the details in private

File: counsel.el [hunk 1]
```diff
@@ -1919,25 +1919,38 @@ INITIAL-INPUT can be given as the initial minibuffer input."
...
+    (counsel-require-program fzf-basename)
+    (setq counsel--fzf-dir
+          (if (and
+               (fboundp 'projectile-project-p)
+               (fboundp 'projectile-project-root)
+               (projectile-project-p))
+              (projectile-project-root)
+            (or initial-directory
+                default-directory)))
+    (ivy-set-prompt 'counsel-ag counsel-prompt-function)
```
Comment (jojojames) [inline counsel.el]: Should this be counsel-fzf?

Comment (woodensquares) [reply counsel.el]: yes it should, not sure how that slipped in, thanks for the review, let me fix it

**Comment by woodensquares** [comment]
actually just realized that @abo-abo had changed some logic, in the same function, let me integrate his changes

**Comment by agsdot** [comment]
Pardon my ignorance here, but what does ctrl-u do @woodensquares ?

**Comment by agsdot** [comment]
Or rather how do I use ctrl-u with counsel-fzf to "to select the starting directory " ? Thanks!

**Comment by woodensquares** [comment]
it would work similarly to say counsel-rg, you would do ctrl-u counsel-fzf, then you pick a directory, press enter, then fzf would start in the directory you chose rather than in the current directory or in the projectile project root.

**Comment by agsdot** [comment]
Interesting, never knew about that feature @woodensquares . I noticed though I had to drop into emacs key bindings (normally, I use evil keybindings).

@jojojames , you initially put in the counsel-fzf function, and I know that you specialize in evil keybindings (e.g. github.com/jojojames/evil-collections ), what would be a good way to use "ctrl u counsel-fzf" with evil?

**Comment by woodensquares** [comment]
we are definitely going offtopic here :), but I think something like this should work for example to bind a "choose the directory first" counsel-rg to <leader>g in evil

```
(evil-leader/set-key
 "g"
 (lambda ()
   (interactive)
   (let ((current-prefix-arg 4))
     (call-interactively #'counsel-rg))))
```

**Comment by jojojames** [comment]
@agsdot Since you probably have C-u bound, you need to bind universal-argument to another key.

```
C-u runs the command universal-argument (found in global-map), which
is an interactive compiled Lisp function in ‘simple.el’.

It is bound to C-u.

(universal-argument)
```

You could also use the \ command to switch to emacs state first before triggering C-u.

```
\ runs the command evil-execute-in-emacs-state (found in
evil-motion-state-map), which is an interactive Lisp function in
‘evil-commands.el’.

It is bound to \.

(evil-execute-in-emacs-state)
```

**Comment by agsdot** [comment]
Interesting, @jojojames .

I'll consider doing that or binding that interactive lambda function to an evil-leader / general keybinding as @woodensquares suggested.

Thanks!

**Comment by abo-abo** [comment]
Merged, thanks.

**Comment by agsdot** [comment]
Thanks @woodensquares @jojojames @abo-abo for developing and refining the counsel-fzf function. Much appreciated.

**Pull request closed without merging**